### PR TITLE
[RLlib] Cleanup examples folder (vol 23): Float16 training support and new example script.

### PR DESCRIPTION
**Pull request by sven1977**
Cleanup examples folder (vol 23): Float16 training support and new example script.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/rllib/package_ref/learner.rst [hunk 1]
```diff
@@ -91,7 +91,7 @@ Computing Losses
     :nosignatures:
     :toctree: doc/
 
-    Learner.compute_loss
+    Learner.compute_losses
```
Comment (sven1977) [inline doc/source/rllib/package_ref/learner.rst:94]: I changed this API name to be even more clear. This method computes one(!) loss per RLModule (in a MultiRLModule) inside the Learner.

Got rid of the confusing `TOTAL_LOSS` key. We compute this now in the default implementation of `compute_gradients`.

File: rllib/algorithms/ppo/torch/ppo_torch_learner.py [hunk 2]
```diff
@@ -98,9 +99,8 @@ def possibly_masked_mean(data_):
             mean_vf_unclipped_loss = possibly_masked_mean(vf_loss)
         # Ignore the value function.
         else:
-            value_fn_out = torch.tensor(0.0).to(surrogate_loss.device)
-            mean_vf_unclipped_loss = torch.tensor(0.0).to(surrogate_loss.device)
-            vf_loss_clipped = mean_vf_loss = torch.tensor(0.0).to(surrogate_loss.device)
+            z = torch.tensor(0.0, device=surrogate_loss.device)
```
Comment (sven1977) [inline rllib/algorithms/ppo/torch/ppo_torch_learner.py:102]: simplify

File: rllib/core/learner/learner.py [hunk 3]
```diff
@@ -1621,3 +1624,13 @@ def get_optimizer_state(self, *args, **kwargs):
     @Deprecated(new="Learner._set_optimizer_state()", error=True)
     def set_optimizer_state(self, *args, **kwargs):
         pass
+
+    @Deprecated(new="Learner.compute_losses(...)", error=False)
```
Comment (sven1977) [inline rllib/core/learner/learner.py:1628]: Emulate the old behavior, in case users use this in their tests or other code.

File: rllib/BUILD [hunk 4]
```diff
@@ -2969,6 +2969,14 @@ py_test(
...
 # subdirectory: gpus/
 # ....................................
+py_test(
+    name = "examples/gpus/float16_training_and_inference",
+    main = "examples/gpus/float16_training_and_inference.py",
+    tags = ["team:rllib", "exclusive", "examples", "gpu"],
+    size = "medium",
+    srcs = ["examples/gpus/float16_training_and_inference.py"],
+    args = ["--enable-new-api-stack", "--as-test", "--stop-reward=150.0"]
+)
```
Comment (simonsays1980) [inline rllib/BUILD:2979]: Interesting. Is this even used by users?

File: rllib/algorithms/algorithm_config.py [hunk 5]
```diff
@@ -3194,6 +3197,14 @@ def experimental(
         """Sets the config's experimental settings.
 
         Args:
+            _torch_grad_scaler_class: Class to use for torch loss scaling (and gradient
+                unscaling). The class must implement the following methods to be
+                compatible with a `TorchLearner`. These methods/APIs match exactly the
```
Comment (simonsays1980) [inline rllib/algorithms/algorithm_config.py:3202]: Small nit: Remove "the" at the end.

File: rllib/algorithms/algorithm_config.py [hunk 6]
```diff
@@ -3194,6 +3197,14 @@ def experimental(
...
+                those of torch's own `torch.amp.GradScaler`:
+                `scale([loss])` to scale the loss.
+                `get_scale()` to get the current scale value.
+                `step([optimizer])` to unscale the grads and step the given optimizer.
+                `update()` to update the scaler after an optimizer step.
```
Comment (simonsays1980) [inline rllib/algorithms/algorithm_config.py:3207]: The description is not clear enough imo. For what is it used?

File: rllib/algorithms/ppo/torch/ppo_torch_learner.py [hunk 7]
```diff
@@ -47,10 +47,11 @@ def compute_loss_for_module(
         # for which we add an additional (artificial) timestep to each episode to
         # simplify the actual computation.
         if Columns.LOSS_MASK in batch:
-            num_valid = torch.sum(batch[Columns.LOSS_MASK])
+            mask = batch[Columns.LOSS_MASK]
+            num_valid = torch.sum(mask)
 
             def possibly_masked_mean(data_):
-                return torch.sum(data_[batch[Columns.LOSS_MASK]]) / num_valid
+                return torch.sum(data_[mask]) / num_valid
```
Comment (simonsays1980) [inline rllib/algorithms/ppo/torch/ppo_torch_learner.py:54]: Could we potentially also mask with this `mask` values in the observations that are not available in a certain step of the environment, e.g. different number of entities at different steps of a game?

File: rllib/algorithms/ppo/torch/ppo_torch_learner.py [hunk 8]
```diff
@@ -98,9 +99,8 @@ def possibly_masked_mean(data_):
             mean_vf_unclipped_loss = possibly_masked_mean(vf_loss)
         # Ignore the value function.
         else:
-            value_fn_out = torch.tensor(0.0).to(surrogate_loss.device)
-            mean_vf_unclipped_loss = torch.tensor(0.0).to(surrogate_loss.device)
-            vf_loss_clipped = mean_vf_loss = torch.tensor(0.0).to(surrogate_loss.device)
+            z = torch.tensor(0.0, device=surrogate_loss.device)
+            value_fn_out = mean_vf_unclipped_loss = vf_loss_clipped = mean_vf_loss = z
```
Comment (simonsays1980) [inline rllib/algorithms/ppo/torch/ppo_torch_learner.py:103]: Niceee!

File: rllib/core/learner/learner.py [hunk 9]
```diff
@@ -214,8 +214,8 @@ class Learner(Checkpointable):
 
             class MyLearner(TorchLearner):
 
-               def compute_loss(self, fwd_out, batch):
-                   # compute the loss based on batch and output of the forward pass
+               def compute_losses(self, fwd_out, batch):
+                   # Compute the loss based on batch and output of the forward pass
                    # to access the learner hyper-parameters use `self._hps`
```
Comment (simonsays1980) [inline rllib/core/learner/learner.py:219]: Remove `_hps`. Hyperparameters are not used anymore.

File: rllib/core/learner/learner.py [hunk 10]
```diff
@@ -830,33 +843,30 @@ def should_module_be_updated(self, module_id, multi_agent_batch=None):
...
-        loss calculations.
+        """Computes the loss(es) for the module being optimized.
+
+        This method must be overridden by MultiRLModule-specific Learners in order to
+        define the specific loss computation logic. If the algorithm is single-agent
+        `compute_loss_for_module()` should be overridden instead. If the algorithm uses
+        independent multi-agent learning (default behavior for multi-agent setups), also
+        `compute_loss_for_module()` should be overridden, but it will be called for each
+        individual RLModule inside the MultiRLModule.
+        It is recommended to not compute any forward passes within this method, and to
```
Comment (simonsays1980) [inline rllib/core/learner/learner.py:857]: Very important hint! Awesome!

File: rllib/core/learner/learner.py [hunk 11]
```diff
@@ -830,33 +843,30 @@ def should_module_be_updated(self, module_id, multi_agent_batch=None):
...
-        `forward_train()` outputs of the RLModule(s) to compute the required tensors for
-        loss calculations.
+        """Computes the loss(es) for the module being optimized.
+
+        This method must be overridden by MultiRLModule-specific Learners in order to
+        define the specific loss computation logic. If the algorithm is single-agent
+        `compute_loss_for_module()` should be overridden instead. If the algorithm uses
+        independent multi-agent learning (default behavior for multi-agent setups), also
+        `compute_loss_for_module()` should be overridden, but it will be called for each
+        individual RLModule inside the MultiRLModule.
```
Comment (simonsays1980) [inline rllib/core/learner/learner.py:856]: Let's either refer to the docs about this or to some examples. Just from the text this does not become very clear.

File: rllib/core/learner/learner.py [hunk 12]
```diff
@@ -893,7 +896,7 @@ def compute_loss_for_module(
 
         Think of this as computing loss for a single agent. For multi-agent use-cases
         that require more complicated computation for loss, consider overriding the
-        `compute_loss` method instead.
+        `compute_losses` method instead.
```
Comment (simonsays1980) [inline rllib/core/learner/learner.py:899]: Can `compute_losses` call internally `compute_loss_for_module` or is the latter called nevertheless and losses would be computed two times?

File: rllib/core/learner/tf/tf_learner.py [hunk 13]
```diff
@@ -300,7 +300,7 @@ def _untraced_update(
         def helper(_batch):
             with tf.GradientTape(persistent=True) as tape:
                 fwd_out = self._module.forward_train(_batch)
-                loss_per_module = self.compute_loss(fwd_out=fwd_out, batch=_batch)
+                loss_per_module = self.compute_losses(fwd_out=fwd_out, batch=_batch)
```
Comment (simonsays1980) [inline rllib/core/learner/tf/tf_learner.py:303]: Do we even want this here and motivate users to write TF algorithms?

File: rllib/core/learner/torch/torch_learner.py [hunk 14]
```diff
@@ -108,7 +117,7 @@ def configure_optimizers_for_module(
 
         # For this default implementation, the learning rate is handled by the
         # attached lr Scheduler (controlled by self.config.lr, which can be a
-        # fixed value of a schedule setting).
+        # fixed value or a schedule setting).
         params = self.get_parameters(module)
         optimizer = torch.optim.Adam(params)
```
Comment (simonsays1980) [inline rllib/core/learner/torch/torch_learner.py:122]: Don't we want to pass in `kwargs` to `Adam`?

File: rllib/core/learner/torch/torch_learner.py [hunk 15]
```diff
@@ -149,24 +158,48 @@ def compute_gradients(
         for optim in self._optimizer_parameters:
             # `set_to_none=True` is a faster way to zero out the gradients.
             optim.zero_grad(set_to_none=True)
-        loss_per_module[ALL_MODULES].backward()
+
+        if self._grad_scalers is not None:
```
Comment (simonsays1980) [inline rllib/core/learner/torch/torch_learner.py:162]: Awesome!!!

File: rllib/examples/gpus/float16_training_and_inference.py [hunk 16]
```diff
@@ -0,0 +1,241 @@
+"""Example of using float16 precision for training and inference.
+
+This example:
+    - shows how to write a custom callback for RLlib to convert all RLModules
```
Comment (simonsays1980) [inline rllib/examples/gpus/float16_training_and_inference.py:4]: Absolutely awesome!!! This is such a great example!

**Review by simonsays1980 — APPROVED** [review]
LGTM. Awesome PR. It would be nice to add some more documentation of why certain components are needed and how the new loss setup works now.

File: rllib/BUILD (hunk 4 above)
Comment (sven1977) [reply rllib/BUILD:2979]: Yes. Actually a customer asked for it ;)
It does get interesting for large models (see also many efforts training LLMs with super compressed precisions down to bfloat16) and multi-agent. I think if one can stabilize this, it's very useful. Another example script with mixed-precision training (and float16 inference on the EnvRunners) is in the making ...

File: rllib/algorithms/algorithm_config.py (hunk 5 above)
Comment (sven1977) [reply rllib/algorithms/algorithm_config.py:3202]: done

File: rllib/algorithms/algorithm_config.py (hunk 6 above)
Comment (sven1977) [reply rllib/algorithms/algorithm_config.py:3207]: enhanced and added a link to the torch docs

File: rllib/algorithms/ppo/torch/ppo_torch_learner.py (hunk 7 above)
Comment (sven1977) [reply rllib/algorithms/ppo/torch/ppo_torch_learner.py:54]: That should work. But you would have to manually change that in the Learner connector (where this column is being produced and added to the batch).

File: rllib/core/learner/learner.py (hunk 9 above)
Comment (sven1977) [reply rllib/core/learner/learner.py:219]: Great catch! Will fix.

Comment (sven1977) [reply rllib/core/learner/learner.py:219]: fixed

File: rllib/core/learner/learner.py (hunk 11 above)
Comment (sven1977) [reply rllib/core/learner/learner.py:856]: I can link to the custom loss example ...

Comment (sven1977) [reply rllib/core/learner/learner.py:856]: done

File: rllib/core/learner/tf/tf_learner.py (hunk 13 above)
Comment (sven1977) [reply rllib/core/learner/tf/tf_learner.py:303]: Good point! But .... DreamerV3 :|

File: rllib/core/learner/learner.py (hunk 12 above)
Comment (sven1977) [reply rllib/core/learner/learner.py:899]: The default implementation of `compute_losses` calls n times `compute_loss_for_module`, where n is the number of RLModules within the Learner's MultiRLModule. So nothing is computed twice.

If you have a complex multi-agent case, you should override `compute_losses`, in which case the n calls to `compute_loss_for_module` will NOT be made.

File: rllib/core/learner/torch/torch_learner.py (hunk 14 above)
Comment (sven1977) [reply rllib/core/learner/torch/torch_learner.py:122]: Used to think the same way.

BUT that would make configuring custom optimizers again very non-pythonic and very yaml'ish. The user would then have to provide a class/type and some kwargs, but then has no chance to customize anything else within the optimizer setup process. Think of a user having to configure two optimizers, or three. Where do you make these options available in the config, then? What if the user needs different optimizers per module?

File: rllib/examples/gpus/float16_training_and_inference.py (hunk 16 above)
Comment (sven1977) [reply rllib/examples/gpus/float16_training_and_inference.py:4]: Thanks a ton!! Yes, and it actually works :D (only with adding all these weird tricks to stabilize learning). But I agree, it's super nice to see how users can customize not just one component at the same time, but several and then achieve a very high degree of customizability w/o us having to add feature after feature to the lib.

**Comment by can-anyscale** [comment]
Broke https://github.com/ray-project/ray/issues/45088 and is blocking the release, i'm reverting to unblock
